src/main/java/org/candlepin/auth/CloudRegistrationAuth.java

```diff
@@ -267,23 +275,142 @@ public String generateRegistrationToken(Principal principal, CloudRegistrationIn
+            throw new IllegalArgumentException("cloudRegistrationInfo is null");
+        }
+
+        CloudAuthenticationResult result = this.cloudRegistrationAdapter
+            .resolveCloudRegistrationDataV2(cloudRegInfo);
+
+        String cloudAccountId = result.getCloudAccountId();
+        if (cloudAccountId == null || cloudAccountId.isBlank()) {
+            String errmsg = this.i18nProvider.get()
+                .tr("cloud provider account could not be resolved");
+
+            throw new CloudRegistrationAuthorizationException(errmsg);
+        }
+
+        String cloudInstanceId = result.getCloudInstanceId();
+        if (cloudInstanceId == null || cloudInstanceId.isBlank()) {
+            String errmsg = this.i18nProvider.get()
+                .tr("cloud instance ID could not be resolved");
+
+            throw new CloudRegistrationAuthorizationException(errmsg);
+        }
+
+        String cloudProviderShortName = result.getCloudProviderShortName();
+        if (cloudProviderShortName == null || cloudProviderShortName.isBlank()) {
+            String errmsg = this.i18nProvider.get()
+                .tr("cloud provider short name could not be resolved");
+
+            throw new CloudRegistrationAuthorizationException(errmsg);
+        }
+
+        String productId = result.getProductId();
+        if (productId == null || productId.isBlank()) {
+            String errmsg = this.i18nProvider.get()
+                .tr("product id could not be resolved");
+
+            throw new CloudRegistrationAuthorizationException(errmsg);
+        }
+
+        String ownerKey = result.getOwnerKey();
+        boolean ownerAndPoolExists = false;
+        if (ownerKey != null && !ownerKey.isBlank() && result.isEntitled()) {
+            Owner owner = ownerCurator.getByKey(ownerKey);
+            if (owner != null) {
+                ownerAndPoolExists = !poolCurator.listByOwnerAndProduct(owner, productId).isEmpty();
+            }
+        }
+
+        String anonymousConsumerUuid = null;
+        if (ownerKey == null || ownerKey.isBlank() || !result.isEntitled() || !ownerAndPoolExists) {
+
+            // TODO: Schedule anonymous org creation job
+
+            AnonymousCloudConsumer anonConsumer = anonymousCloudConsumerCurator
+                .getByCloudInstanceId(cloudInstanceId);
+
+            if (anonConsumer == null) {
+                anonConsumer = new AnonymousCloudConsumer()
+                    .setCloudAccountId(cloudAccountId)
+                    .setCloudInstanceId(cloudInstanceId)
+                    .setProductId(productId)
+                    .setCloudProviderShortName(cloudProviderShortName);
+
+                anonConsumer = anonymousCloudConsumerCurator.create(anonConsumer);
+            }
+
+            anonymousConsumerUuid = anonConsumer.getUuid();
+        }
+
+        String tokenType = anonymousConsumerUuid == null ? TOKEN_TYPE : ANON_TOKEN_TYPE;
+        String token = anonymousConsumerUuid == null ?
```

Comment:
If we're introducing a new token type, we have to validate/verify it separately than the current one. It's inevitable. For example in addition to validating the token type itself, we'd have to check if we actually have an AnonymousCloudConsumer with the uuid specified in it. I assume this change is not added yet because the PR is still a draft

Reply:
Actually, now I remember why that code isn't added yet. It's because it's part of the next jira card

Reply:
I think splitting up the token creation into two different methods makes sense. One for the regular JWT token and another for the anonymous JWT token. I also think it makes sense to put the different token types into an enum!

Reply:
Yeah, but I feel like we should probably go one step further: a completely separate auth class (`AnonymousRegistrationAuth`?). There's a lot of overlap, but where the deviations occur make it very difficult to properly utilize all the same paths without getting into a case where we're constantly checking the token type, what "mode" we're in, or some other state validation and hand-off.

Reply:
My bad on this, I misunderstood what you were suggesting on separating. I agree that it makes sense to seperate the auth for the anonymous token type into a new class!